scripts/coverage_report_jenkins.py

```diff
@@ -51,6 +118,224 @@ def get_build_numbers(client, job_name):
+
+    Returns:
+        Nothing
+    """
+    print('Merging coverage data')
+
+    # Run the coverage merger script out of the rails root pointing
+    # to where the coverage data is installed.   This will generate
+    # a directory under the coverage directory called merged, and
+    # will have the merged .resultset.json file in it, along with some
+    # HTML that was generated by the merger script.
+    cmd = ssh.run_rails_command(
+        'coverage_merger.rb --coverageRoot={}'.format(coverage_dir),
+        timeout=60 * 60)
+    if not cmd:
+        print('Failure running the merger - {}'.format(str(cmd)))
+        return 6
+    else:
+        print('Coverage report generation was successful')
+        print(str(cmd))
+        percentage = re.search(r'LOC\s+\((\d+.\d+%)\)\s+covered\.', str(cmd))
+        if percentage:
+            print('COVERAGE={};'.format(percentage.groups()[0]))
+        else:
+            print('COVERAGE=unknown;')
+
+    # The sonar-scanner will actually need the .resultset.json file it
+    # uses to be in /coverage/.resultset.json (i.e. the root of the coverarage
+    # directory), so lets create a symlink:
+    ssh.run_command('ln -s merged/.resultset.json {}/.resultset.json'.format(coverage_dir))
+
+
+def pull_merged_coverage_data(ssh, coverage_dir):
+    """Pulls merged coverage data to log directory.
+
+    Args:
+        ssh:  ssh client
+        coverage_dir:  Directory where the coverage archive was extracted.
+
+    Returns:
+        Nothing
+    """
+    print('Packing the generated HTML')
+    cmd = ssh.run_command('cd {}; tar cfz /tmp/merged.tgz merged'.format(coverage_dir))
+    if not cmd:
+        print('Could not compress! - {}'.format(str(cmd)))
+        return 7
+    print('Grabbing the generated HTML')
+    ssh.get_file('/tmp/merged.tgz', log_path.strpath)
+    print('Decompressing the generated HTML')
+    rc = subprocess.call(
+        ['tar', 'xf', log_path.join('merged.tgz').strpath, '-C', log_path.strpath])
+    if rc == 0:
+        print('Done!')
+    else:
+        print('Failure to extract')
+        return 8
+
+
+def install_sonar_scanner(appliance, scanner_url, scanner_dir, server_url):
+    """ Install sonar-scanner application
+
+    Pulls the sonar-scanner application to the appliance from scanner_url,
+    installs it in scanner_dir, and configures it to send its scan data to
+    server_url.  It also configures the project config for the scan, setting
+    sonar.projectVersion to the appliance version, and setting sonar.sources
+    to pick up both sets of sources.
+
+    Args:
+        appliance: Appliance object
+        scanner_url:  Where to get the scanner from.
+        scanner_dir:  Where to install the scanner on the appliance.
+        server_url:  Where to send scan data to (i.e. what sonarqube)
+
+    Returns:
+        Nothing
+    """
+    print "Installing sonar scanner on appliance."
+    ssh = appliance.ssh_client
```

Comment:
This should be in a context manager, so that the ssh_client instance is closed properly.

Reply:
I'm going pass the appliance attributes I need individually because the ssh object is already in a context manager in called code as per our discussion.